### PR TITLE
fix(external-authentication): scope role-deletion impact to the role's tenant

### LINKED ISSUE #8013: External-auth role-deletion impact is not tenant-filtered
## Problem

`ExternalAuthenticationRoleDeletionDependencyContributor` scans connections without an explicit tenant filter when computing role-deletion impact.

## Evidence

Relevant source: `src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs`.

If identical role IDs exist in multiple tenants, impact for one tenant can include references owned by another tenant.

## Expected behavior

Deletion impact and remediation enumerate only dependencies in the role's exact tenant context.

## Context

Discovered while implementing elsa-workflows/elsa-studio#960. This is separate Core work and does not authorize changing the Studio contract.

## Validation

Add cross-tenant impact/remediation tests proving dependencies cannot cross tenant boundaries.

### PR DESCRIPTION
**Pull request by sfmskywalker**
## Purpose

Make the external-authentication role-deletion contributor enumerate and remediate only the connections in the role's own tenant context, so a role ID that exists in several tenants cannot pull another tenant's JIT-policy references into impact or remediation.

## Scope

- [x] Bug fix (behavior change)

### Problem

`ExternalAuthenticationRoleDeletionDependencyContributor.InspectAsync` scanned every stored connection with an empty filter, and the validation and remediation paths loaded connections by owner ID with no tenant check. With identical role IDs in two tenants, tenant A's deletion impact could list tenant B's connection, and a caller-supplied owner ID could reach tenant B's connection during remediation.

### Solution

- The boundary is the resolved role's own tenant, not the ambient tenant: the contributor resolves the role once per operation and scopes to `Role.TenantId` plus host scope, falling back to the ambient tenant only when the role cannot be resolved. That matters on the EF Core path with multitenancy disabled, where no tenant query filter exists and a tenant-owned role can be resolved and deleted by ID. Impact reads the connection store once, as a single snapshot, and filters it in memory to that context. One read means a connection whose tenant changes mid-operation cannot fall between two reads. Configuration-sourced connections are filtered by the same rule while keeping their original index, since the index is part of the path an operator edits.
- Validation and remediation load each dependency through a helper that treats a connection outside the role's tenant scope as absent, so the existing `connection_revision_changed` conflict path fails the request instead of mutating across the boundary. No new result codes; the `IRoleDeletionDependencyContributor` contract and Studio-facing models are unchanged.
- Host-scoped connections (`"*"`) stay in scope for every tenant. That rule comes from the provisioning code: the connection registry serves the host scope to every signing-in tenant, and a connection's default role IDs are resolved through the tenant-filtered role store of the signing-in user, so a host connection naming role X does reference tenant A's role X. Documented in the contributor's XML remarks.
- Tenant-agnostic roles (`TenantId == "*"`) are visible from every tenant, so their tenant context is every tenant: for such a role the contributor scans every stored connection and every configuration entry regardless of tenant, and remediation may update any of them. The role is resolved once per operation through the active role store. Durable persistence keys roles by ID alone (`PK_Roles`), so the lookup is exact. Only the in-memory store can hold two roles with the same ID; that ambiguous case fails closed with an `InvalidOperationException`, because widening would expose other tenants' references for a tenant-scoped deletion and narrowing would leave an agnostic role's references dangling. A role that cannot be resolved keeps the tenant-scoped behavior. When remediating an agnostic role requires a replacement default role, the replacement must itself be agnostic, since only a role that exists identically in every tenant can be written into another tenant's policy safely; otherwise the existing `replacement_role_unavailable_or_unauthorized` code is returned. A replacement ID that resolves to more than one role is rejected with the same code rather than guessed at, and the check is repeated at removal time.
- The contributor now takes `ITenantAccessor`; `AddExternalAuthenticationServices` registers a default accessor with `TryAdd`, mirroring the identity feature, so the module still resolves on a host that has not enabled multitenancy.

### Tests

Unit tests cover: impact excludes a same-ID reference owned by another tenant; impact includes host-scoped connections for every tenant; remediation cannot reach another tenant's connection even when its owner ID is supplied; remediation stops if the connection leaves the role's tenant after validation; remediation still removes the role from a host-scoped connection. Three further tests cover an agnostic role: impact includes a connection owned by another tenant, remediation removes the reference from it, and a role ID that resolves to more than one role fails closed in inspection, validation, and remediation. An EF Core integration test pins that the connection store honors `ConnectionFilter.Scope`. With the tenant scoping reverted, three of the new tests fail.

## Verification

- `dotnet test test/unit/Elsa.ExternalAuthentication.UnitTests/...`: 202 passed.
- `dotnet test test/integration/Elsa.ExternalAuthentication.IntegrationTests/...`: 150 passed.
- `dotnet test test/unit/Elsa.Identity.UnitTests/...`: 131 passed.
- `dotnet build Elsa.sln`: 0 errors; warnings pre-exist on main and are outside the touched files.

Noted and left for a separate decision: a tenant administrator can still remediate a host-scoped connection and write a tenant-scoped replacement into it, which matches the module's existing authorization model for host connections. Requiring agnostic replacements there would block every replacement in a deployment without multitenancy, so the correct rule depends on whether multitenancy is active.

Fixes #8013

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 1]
```diff
@@ -191,6 +206,48 @@ await notifier.PublishAsync(
...
+    private async ValueTask<IdentityProviderConnection?> FindConnectionInRoleTenantScopeAsync(string ownerId, CancellationToken cancellationToken)
+    {
+        var connection = await store.FindByIdAsync(ownerId, cancellationToken);
+        return connection is not null && IsInRoleTenantScope(connection.TenantId) ? connection : null;
+    }
+
+    private IReadOnlyCollection<ConnectionScope> GetRoleTenantScopes()
+    {
+        var tenantScope = ToScope(tenantAccessor.TenantId.NormalizeTenantId());
+        return tenantScope == ConnectionScope.Host ? [ConnectionScope.Host] : [ConnectionScope.Host, tenantScope];
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs]: Addressed on current head 171aeb7. The contributor now resolves the role being deleted once per operation through the active role store; when its TenantId is Tenant.AgnosticTenantId the tenant context is every tenant, so InspectAsync scans every stored connection and every configuration entry unfiltered, and ValidateRemovalAsync/RemoveEditableReferencesAsync accept any connection loaded by owner ID. Named-tenant roles keep the scoped behavior. Covered by ImpactForAnAgnosticRoleIncludesAConnectionOwnedByAnotherTenant and RemediationOfAnAgnosticRoleCanRemoveTheReferenceFromAnotherTenantsConnection, both verified to fail without the agnostic branch.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 2]
```diff
@@ -191,6 +216,69 @@ await notifier.PublishAsync(
...
+    /// </summary>
+    private async ValueTask<IReadOnlyCollection<IdentityProviderConnection>> FindConnectionsInRoleTenantScopeAsync(bool isAgnosticRole, CancellationToken cancellationToken)
+    {
+        if (isAgnosticRole)
+            return (await store.FindAsync(new(), cancellationToken)).Items.ToArray();
+
+        var connections = new List<IdentityProviderConnection>();
+        foreach (var scope in GetRoleTenantScopes())
+            connections.AddRange((await store.FindAsync(new()
+                { Scope = scope }, cancellationToken)).Items);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs]: Addressed on current head 8956c3a (and later). FindConnectionsInRoleTenantScopeAsync now performs a single store.FindAsync(new(), ct) and filters that one snapshot in memory with IsInRoleTenantScope for both the tenant-scoped and the agnostic case, so a connection whose TenantId changes mid-operation has no gap between reads to fall through. The per-scope helpers were removed.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 3]
```diff
@@ -191,6 +216,69 @@ await notifier.PublishAsync(
...
+    /// which case its tenant context is every tenant rather than the ambient one. Resolves the role through the
+    /// active role store exactly the way <see cref="RemoveEditableReferencesAsync"/> already does for a
+    /// replacement role. A missing store or an unresolved role keeps the current tenant-scoped behavior.
+    /// </summary>
+    private async ValueTask<bool> IsAgnosticRoleAsync(string roleId, CancellationToken cancellationToken)
+    {
+        var roleStore = ActiveRoleStore;
+        if (roleStore is null)
+            return false;
+        var role = await roleStore.FindAsync(new() { Id = roleId }, cancellationToken);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs]: Addressed on current head 8956c3a. IsAgnosticRoleAsync now uses IRoleStore.FindManyAsync(new() { Id = roleId }) and treats the deletion as agnostic when any role sharing the ID is agnostic, so the wider every-tenant scan wins an ID collision deterministically. RoleFilter.Apply and MemoryRoleStore.Filter both return the agnostic row alongside the ambient tenant's row, which is what makes this reliable in both stores. Covered by ImpactForAnAgnosticRoleThatSharesAnIdWithTheAmbientTenantsRoleIncludesAConnectionOwnedByAnotherTenant, which seeds the tenant role first and fails against the previous first-match lookup.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 4]
```diff
@@ -191,6 +216,59 @@ await notifier.PublishAsync(
...
+    /// agnostic: over-scanning at worst surfaces a reference the operator can inspect, where under-scanning
+    /// would leave one dangling. A missing store or an unresolved role keeps the current tenant-scoped behavior.
+    /// </summary>
+    private async ValueTask<bool> IsAgnosticRoleAsync(string roleId, CancellationToken cancellationToken)
+    {
+        var roleStore = ActiveRoleStore;
+        if (roleStore is null)
+            return false;
+        var roles = await roleStore.FindManyAsync(new() { Id = roleId }, cancellationToken);
+        return roles.Any(role => string.Equals(role.TenantId, Tenant.AgnosticTenantId, StringComparison.Ordinal));
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs]: Addressed on the current head. The contributor no longer widens on any same-ID match. It resolves the role once per operation: with exactly one match it uses that role's tenant (agnostic or not); with more than one match it throws InvalidOperationException, which the coordinator does not catch, so the deletion request fails closed instead of guessing in either direction. Two facts make this the right shape: durable persistence keys roles by Id alone (PK_Roles in every provider's Identity Initial migration), so the collision cannot exist there and the lookup is exact; and only MemoryRoleStore can hold the collision, where the coordinator's own delete is already ambiguous, so no contributor-side guess could make the operation consistent. Covered by RoleIdThatResolvesToMoreThanOneRoleAcrossTenantScopesFailsClosed for Inspect, Validate, and Remove.

[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs]: Addressed on the current head. The contributor no longer widens on any same-ID match. It resolves the role once per operation: with exactly one match it uses that role's tenant (agnostic or not); with more than one match it throws InvalidOperationException, which the coordinator does not catch, so the deletion request fails closed instead of guessing in either direction. Two facts make this the right shape: durable persistence keys roles by Id alone (PK_Roles in every provider's Identity Initial migration), so the collision cannot exist there and the lookup is exact; and only MemoryRoleStore can hold the collision, where the coordinator's own delete is already ambiguous, so no contributor-side guess could make the operation consistent. Covered by RoleIdThatResolvesToMoreThanOneRoleAcrossTenantScopesFailsClosed for Inspect, Validate, and Remove.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 5]
```diff
@@ -25,24 +46,34 @@ public sealed class ExternalAuthenticationRoleDeletionDependencyContributor(
...
 
+    /// <summary>
+    /// Match the default DI container's direct-service semantics: when persistence replaces the in-memory
+    /// store, the last registration is the active store.
+    /// </summary>
+    private IRoleStore? ActiveRoleStore => roleStores.LastOrDefault();
+
     public async ValueTask<RoleDeletionDependencySnapshot> InspectAsync(string roleId, CancellationToken cancellationToken = default)
     {
+        var isAgnosticRole = await IsAgnosticRoleAsync(roleId, cancellationToken);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs]: Addressed on the current head. The resolution now returns the resolved role's own normalized TenantId (agnostic, or a concrete tenant), and that value is threaded through inspection, validation, and remediation in place of the boolean; the ambient tenant is read only in the documented unresolved-role fallback. Covered by ImpactIsScopedByTheResolvedRolesTenantRatherThanTheAmbientTenant, which deletes a tenant-A role while the default tenant is ambient through a store that resolves by ID without an ambient filter (the EF-without-multitenancy shape) and asserts tenant A's reference is included and tenant B's excluded; it fails without the change.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 6]
```diff
@@ -25,24 +50,34 @@ public sealed class ExternalAuthenticationRoleDeletionDependencyContributor(
     IConnectionRegistryVersionStore registryVersions,
     ConnectionRevisionCalculator revisionCalculator,
     ExternalAuthenticationSecurityNotifier notifier,
-    IPermissionEvaluator permissionEvaluator) : IRoleDeletionDependencyContributor
+    IPermissionEvaluator permissionEvaluator,
+    ITenantAccessor tenantAccessor) : IRoleDeletionDependencyContributor
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs:61]: Keeping the single constructor. ExternalAuthenticationRoleDeletionDependencyContributor was introduced on 2026-07-25 (95b2ce8d8) and is absent from the latest release, 3.8.0, so no published consumer can instantiate it; on main it is constructed only by the module's DI registration and by its own unit tests, which this PR updates. A compatibility overload with a fresh DefaultTenantAccessor would make the fallback path silently depend on an accessor no host controls, which is the kind of implicit tenant context this fix removes. Recording this as a deliberate won't-fix on the PR.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 7]
```diff
@@ -191,6 +226,68 @@ await notifier.PublishAsync(
...
+    private async ValueTask<IdentityProviderConnection?> FindConnectionInRoleTenantScopeAsync(string ownerId, string roleTenantId, CancellationToken cancellationToken)
+    {
+        var connection = await store.FindByIdAsync(ownerId, cancellationToken);
+        return connection is not null && IsInRoleTenantScope(connection.TenantId, roleTenantId) ? connection : null;
+    }
+
+    private static bool IsInRoleTenantScope(string? connectionTenantId, string roleTenantId) =>
+        string.Equals(roleTenantId, Tenant.AgnosticTenantId, StringComparison.Ordinal) ||
+        string.Equals(connectionTenantId, ConnectionScope.HostTenantId, StringComparison.Ordinal) ||
+        string.Equals(connectionTenantId.NormalizeTenantId(), roleTenantId, StringComparison.Ordinal);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs:340]: Addressed on the current head for the part that grants something new: when the deletion target is agnostic and remediation needs a replacement default role, the replacement must itself resolve to an agnostic role, otherwise validation returns Forbidden and removal returns Failed with the existing replacement_role_unavailable_or_unauthorized code and the other tenant's connection is untouched. Covered by RemediationOfAnAgnosticRoleRejectsATenantScopedReplacementRole (fails without the guard) and RemediationOfAnAgnosticRoleAcceptsAnAgnosticReplacementRole. The remaining-role authorization runs in the ambient tenant as before: those roles are already present on the policy and are not newly granted by remediation, and that is the module's existing authorization model for host-scoped connections, which every tenant may already remediate. Per-connection-tenant authorization of remaining roles is a broader change to that model than this issue authorizes; I have noted it on the PR as a follow-up candidate.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 8]
```diff
@@ -104,6 +143,14 @@ public async ValueTask<RoleReferenceRemovalValidationResult> ValidateRemovalAsyn
                  string.Equals(request.ReplacementRoleId, request.RoleId, StringComparison.Ordinal)))
                 return new RoleReferenceRemovalValidationResult.Forbidden("replacement_role_unavailable_or_unauthorized");
 
+            // Authorization below still resolves through the ambient tenant's role services, so an agnostic
+            // deletion target may only be replaced by another agnostic role; a tenant-scoped replacement would
+            // otherwise be authorized in this tenant and then written into every other tenant's connections.
+            if (requiresReplacement &&
+                string.Equals(roleTenantId, Tenant.AgnosticTenantId, StringComparison.Ordinal) &&
+                !await IsAgnosticRoleAsync(request.ReplacementRoleId, cancellationToken))
+                return new RoleReferenceRemovalValidationResult.Forbidden("replacement_role_unavailable_or_unauthorized");
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs:161]: Not applying this one, after trying it. IdentityProviderConnectionManagementService forces every managed connection to host scope, and a deployment without multitenancy creates roles scoped to the default tenant rather than agnostic, so requiring an agnostic replacement for host connections would make every replacement remediation impossible in the default deployment (three existing tests broke for exactly that reason). Writing a tenant-A replacement into a host connection is the module's pre-existing behavior for host connections and is outside what #8013 authorizes; the rule that would be correct depends on whether multitenancy is active, which is a broader change to the connection authorization model. Recording as a deliberate won't-fix; the agnostic-deletion-target guard stays.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 9]
```diff
@@ -147,15 +195,21 @@ public async ValueTask<RoleReferenceRemovalResult> RemoveEditableReferencesAsync
...
                         !await roleAuthorizationService.CanAssignRolesAsync(request.Actor, [replacement.Id], cancellationToken))
                         return new RoleReferenceRemovalResult.Failed("replacement_role_unavailable_or_unauthorized", changedOwnerIds);
+
+                    // Authorization above still resolves through the ambient tenant's role services, so an
+                    // agnostic deletion target may only be replaced by another agnostic role; a tenant-scoped
+                    // replacement would otherwise be authorized in this tenant and then written into every other
+                    // tenant's connections.
+                    if (string.Equals(roleTenantId, Tenant.AgnosticTenantId, StringComparison.Ordinal) &&
+                        !string.Equals(replacement.TenantId.NormalizeTenantId(), Tenant.AgnosticTenantId, StringComparison.Ordinal))
+                        return new RoleReferenceRemovalResult.Failed("replacement_role_unavailable_or_unauthorized", changedOwnerIds);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs:226]: Addressed on the current head. The removal-time recheck now calls the same IsAgnosticRoleAsync used at validation, which requires exactly one matching role and that it be agnostic, so a same-ID role added between validation and mutation makes the replacement ambiguous and the removal returns Failed with replacement_role_unavailable_or_unauthorized before any connection is updated.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 10]
```diff
@@ -191,6 +245,82 @@ await notifier.PublishAsync(
...
+    /// same resolution as <see cref="ResolveRoleTenantIdAsync"/>. A role ID that cannot be resolved to exactly
+    /// one role is treated as not agnostic, since there is then no resolved role to trust as safe to write into
+    /// every tenant's connections.
+    /// </summary>
+    private async ValueTask<bool> IsAgnosticRoleAsync(string? roleId, CancellationToken cancellationToken)
+    {
+        if (string.IsNullOrWhiteSpace(roleId))
+            return false;
+        var tenantId = await ResolveRoleTenantIdAsync(roleId, cancellationToken);
+        return string.Equals(tenantId, Tenant.AgnosticTenantId, StringComparison.Ordinal);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs]: Addressed on the current head. IsAgnosticRoleAsync resolves the candidate directly through the shared FindRolesByIdAsync lookup and returns true only when exactly one role matches and it is agnostic; zero or more than one matches return false, so validation returns replacement_role_unavailable_or_unauthorized instead of escaping with an exception. Covered by ReplacementRoleIdThatResolvesToBothATenantRoleAndAnAgnosticRoleIsRejectedRatherThanThrowing, which fails without the change. The deletion target keeps its fail-closed exception.

File: src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs [hunk 11]
```diff
@@ -191,6 +259,90 @@ await notifier.PublishAsync(
...
+        return roles.Length == 1 && string.Equals(roles[0].TenantId.NormalizeTenantId(), Tenant.AgnosticTenantId, StringComparison.Ordinal);
+    }
+
+    /// <summary>Loads every role matching the given ID from the active role store, or an empty result if none is configured.</summary>
+    private async ValueTask<Role[]> FindRolesByIdAsync(string roleId, CancellationToken cancellationToken)
+    {
+        var roleStore = ActiveRoleStore;
+        if (roleStore is null)
+            return [];
+        return (await roleStore.FindManyAsync(new() { Id = roleId }, cancellationToken)).ToArray();
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.ExternalAuthentication/Services/ExternalAuthenticationRoleDeletionDependencyContributor.cs:310]: Not fixable inside this contributor, and tracked separately as #8037. The collision you describe can only exist in MemoryRoleStore, whose storage key includes the tenant; every durable provider keys Roles by Id alone (PK_Roles), so a same-ID role in another tenant cannot exist there and the lookup is exact. IRoleStore offers no unscoped query, so the contributor cannot see a non-ambient tenant's same-ID role, and RoleDeletionCoordinator is equally blind in that store: it deletes whichever match the store returns first. The right fix is for MemoryRoleStore to enforce the same uniqueness as durable persistence, which removes the whole class of collisions for the coordinator and every contributor at once; that is #8037. Recording this as a deliberate won't-fix on the PR.
